test/blackbox/common/DDSBlackboxTestsDataReader.cpp

```diff
@@ -535,6 +535,306 @@ TEST(DDSDataReader, datareader_qos_use_topic_qos)
     ASSERT_EQ(control_qos, test_qos);
 }
 
+bool validate_publication_builtin_topic_data(
+        const eprosima::fastdds::dds::builtin::PublicationBuiltinTopicData& pubdata,
+        const eprosima::fastdds::dds::DataWriter& datawriter)
+{
+    bool ret = true;
+
+    auto dw_qos = datawriter.get_qos();
+    auto pub_qos = datawriter.get_publisher()->get_qos();
+
+    eprosima::fastdds::dds::builtin::BuiltinTopicKey_t dw_key, part_key;
+    GuidPrefix_t guid_prefix = datawriter.get_publisher()->get_participant()->guid().guidPrefix;
+
+    // This conversions may be included later in utils
+    dw_key.value[0] = 0;
+    dw_key.value[1] = 0;
+    dw_key.value[2] = static_cast<uint32_t>(datawriter.guid().entityId.value[0]) << 24
+            | static_cast<uint32_t>(datawriter.guid().entityId.value[1]) << 16
+            | static_cast<uint32_t>(datawriter.guid().entityId.value[2]) << 8
+            | static_cast<uint32_t>(datawriter.guid().entityId.value[3]);
+
+    part_key.value[0] = static_cast<uint32_t>(guid_prefix.value[0]) << 24
+            | static_cast<uint32_t>(guid_prefix.value[1]) << 16
+            | static_cast<uint32_t>(guid_prefix.value[2]) << 8
+            | static_cast<uint32_t>(guid_prefix.value[3]);
+    part_key.value[1] = static_cast<uint32_t>(guid_prefix.value[4]) << 24
+            | static_cast<uint32_t>(guid_prefix.value[5]) << 16
+            | static_cast<uint32_t>(guid_prefix.value[6]) << 8
+            | static_cast<uint32_t>(guid_prefix.value[7]);
+    part_key.value[2] = static_cast<uint32_t>(guid_prefix.value[8]) << 24
+            | static_cast<uint32_t>(guid_prefix.value[9]) << 16
+            | static_cast<uint32_t>(guid_prefix.value[10]) << 8
+            | static_cast<uint32_t>(guid_prefix.value[11]);
+
+    ret &= (0 == memcmp(pubdata.key.value, dw_key.value, sizeof(eprosima::fastdds::dds::builtin::BuiltinTopicKey_t)));
+    ret &=
+            (0 ==
+            memcmp(pubdata.participant_key.value, part_key.value,
+            sizeof(eprosima::fastdds::dds::builtin::BuiltinTopicKey_t)));
```

Comment:
Encapsulate in an auxiliary function in `BlackboxTests.hpp` and use it here and in writer's file.